File: java/src/main/java/io/github/muthuishere/toolnexus/agents/Handle.java
package io.github.muthuishere.toolnexus.agents;

import io.github.muthuishere.toolnexus.LlmClient;
import io.github.muthuishere.toolnexus.Request;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;
import java.util.Locale;
import java.util.concurrent.CompletableFuture;

/**
 * A live agent (SPEC §7D "Handle state machine"): {@code {id, def, state, inbox, budget,
 * children}}. The inbox is agent STATE — persistable and observable — never a runtime/language
 * mailbox. Ids are deterministic and parent-scoped ({@code root/coordinator.1/explore.2}), never
 * random. All mutable fields are guarded by the runtime's single lock; {@link #state} is volatile
 * so observers can read it without taking the lock.
 *
 * <p>Handles are capabilities: post/wake what you hold, wait only on handles you spawned.
 */
public final class Handle {

    public enum State {
        IDLE, RUNNING, SUSPENDED, CLOSED;

        /** Lowercase form used in transition traces ({@code "idle"} etc.) — parity with the
         * shared fixtures' state strings. */
        public String label() { return name().toLowerCase(Locale.ROOT); }
    }

    /** "Unlimited" pool sentinel (a quarter of {@code Long.MAX_VALUE} never exhausts and never
     * overflows under roll-up subtraction). */
    public static final long NO_LIMIT = Long.MAX_VALUE / 4;

    public volatile State state = State.IDLE;
    public final Deque<InboxItem> inbox = new ArrayDeque<>();
    public final List<Handle> children = new ArrayList<>();

    // §7D budgets: pools + effective caps — effective = min(own, parent remaining) at spawn.
    public long poolTokens;
    public long poolToolCalls;
    public final long poolChildren;
    public final int effMaxTurns;
    public final int effMaxConcurrent;
    public final int effMaxDepth;
    /** Wall budget (ms); {@code NO_LIMIT} = unbounded. The window opens at the first turn. */
    public final long effMaxWallMs;
    /** Clock millis at the first turn; 0 = not started (wall budget window not yet open). */
    long wallStartMs;

    public long usageTotal;
    public int turnsTotal;
    public Request pendingReq;
    public List<String> pendingPath;
    /** §7D cancellation: the token bound to the in-flight Run ({@code interrupt} cancels it). */
    LlmClient.CancelToken runCancel;
    /** {@code "interrupted" | "closed"} — set BEFORE the token is cancelled, so the abort is
     * classified by state, never by exception type. */
    String abortReason;
    final List<CompletableFuture<TaskResult>> waiters = new ArrayList<>();
    /** Queued wake prompts (per-parent concurrency gate; FIFO slot transfer). */
    final Deque<String> wakeQueue = new ArrayDeque<>();
    /** Items consumed by the in-flight turn — restored on abort (transactional drain). */
    List<InboxItem> drained = List.of();
    /** Set when spawned via the {@code task} tool: {@code agent + ":" + prompt} (reattachment key). */
    String taskKey;
    /** Last completed turn result — what {@code wait} answers immediately on a settled handle. */
    public TaskResult lastResult;
    int runningChildren;
    int seq;

    public final int depth;
    public final String id;
    public final AgentDef def;
    public final Handle parent;

    Handle(String id, AgentDef def, Handle parent) {
        this.id = id;
        this.def = def;
        this.parent = parent;
        this.depth = parent != null ? parent.depth + 1 : 0;
        // §7D carve: effective = min(own, parent remaining).
        Budget own = def.budget != null ? def.budget : new Budget();
        long pTok = parent != null ? parent.poolTokens : NO_LIMIT;
        this.poolTokens = Math.min(own.maxTokens != null ? own.maxTokens : pTok, pTok);
        long pCalls = parent != null ? parent.poolToolCalls : NO_LIMIT;
        this.poolToolCalls = Math.min(own.maxToolCalls != null ? own.maxToolCalls : pCalls, pCalls);
        this.poolChildren = own.maxChildren != null ? own.maxChildren : NO_LIMIT;
        this.effMaxTurns = own.maxTurns != null ? own.maxTurns : 6;
        this.effMaxConcurrent = own.maxConcurrent != null ? own.maxConcurrent : 8;
        this.effMaxDepth = own.maxDepth != null ? own.maxDepth : 3;
        this.effMaxWallMs = own.maxWallMs != null ? own.maxWallMs : NO_LIMIT;
    }
}

File: csharp/src/Toolnexus/Agents/Handle.cs
namespace Toolnexus.Agents;

/// <summary>
/// One live agent (SPEC §7D). State machine:
/// <c>idle → running → (idle | suspended | closed)</c>; <c>suspended → running</c> ONLY via the
/// Answer to its pending Request. The inbox is AGENT STATE (bounded, transactional drain,
/// checkpointable) — never a runtime/language mailbox. Ids are deterministic and parent-scoped
/// (<c>root/coordinator.1/explore.2</c>) — never random.
///
/// Handles are capabilities: post/wake what you hold, wait only on handles you spawned.
/// Handle→handle blocking calls flow strictly rootward; parent→child interaction is non-blocking;
/// downward traversal (close cascade, list, resume) runs from outside the tree.
/// All mutation happens under the runtime's lock; public getters are read-only views.
/// </summary>
public sealed class Handle
{
    public string Id { get; }
    public AgentDef Def { get; internal set; }
    public Handle? Parent { get; }
    public int Depth { get; }

    /// <summary>idle | running | suspended | closed.</summary>
    public string State { get; internal set; } = "idle";

    /// <summary>Remaining token pool (live ledger — usage roll-up drains every ancestor).</summary>
    public long PoolTokens { get; internal set; }

    /// <summary>Remaining tool-call pool.</summary>
    public long PoolToolCalls { get; internal set; }

    /// <summary>Wall-clock deadline (from <c>maxWallMs</c> at spawn, min'd with the parent's).</summary>
    public DateTimeOffset? WallDeadline { get; }

    public long UsageTotal { get; internal set; }
    public int TurnsTotal { get; internal set; }
    public long ToolCallsTotal { get; internal set; }

    /// <summary>The pending Request while suspended (already stamped with this handle's
    /// <c>data.path</c>).</summary>
    public Request? PendingRequest { get; internal set; }

    /// <summary>Last completed turn result — <c>wait</c> resolves with the NEXT result or this
    /// one when the handle is already settled.</summary>
    public AgentResult? LastResult { get; internal set; }

    public int InboxCount { get { lock (Sync) return InboxItems.Count; } }

    public IReadOnlyList<Handle> Children => ChildList;

    /// <summary>Conversation id in the runtime-wide store = the handle id.</summary>
    public string ConvId => Id;

    // ---- runtime-internal state (mutated under the runtime lock only) ----
    internal readonly object Sync;                       // the owning runtime's lock
    internal readonly List<InboxItem> InboxItems = new();
    internal readonly List<Handle> ChildList = new();
    internal readonly List<TaskCompletionSource<AgentResult>> Waiters = new();
    internal readonly List<string> WakeQueue = new();    // queued wake prompts (concurrency gate)
    internal List<InboxItem> Drained = new();            // consumed by the in-flight turn; restored on abort
    internal CancellationTokenSource? Cts;
    internal string? TaskKey;                            // set when spawned via `task` (reattachment)
    internal string? PendingInput;                       // the halted turn's full input (checkpoint)
    internal int? PoolChildren;                          // null = unlimited
    internal int RunningChildren;
    internal int Seq;
    internal readonly int EffMaxTurns;
    internal readonly int EffMaxConcurrent;
    internal readonly int EffMaxDepth;

    internal Handle(string id, AgentDef def, Handle? parent, object sync, DateTimeOffset now)
    {
        Id = id;
        Def = def;
        Parent = parent;
        Sync = sync;
        Depth = parent != null ? parent.Depth + 1 : 0;
        var own = def.Budget ?? new Budget();

        // Carve: effective = min(own, parent remaining) for every pool.
        var pTok = parent?.PoolTokens ?? long.MaxValue;
        PoolTokens = Math.Min(own.MaxTokens ?? pTok, pTok);
        var pCalls = parent?.PoolToolCalls ?? long.MaxValue;
        PoolToolCalls = Math.Min(own.MaxToolCalls ?? pCalls, pCalls);
        var ownDeadline = own.MaxWallMs is long w ? now.AddMilliseconds(w) : (DateTimeOffset?)null;
        WallDeadline = Min(ownDeadline, parent?.WallDeadline);

        PoolChildren = own.MaxChildren;
        EffMaxTurns = own.MaxTurns ?? 6;
        EffMaxConcurrent = own.MaxConcurrent ?? 8;
        EffMaxDepth = own.MaxDepth ?? 3;
    }

    private static DateTimeOffset? Min(DateTimeOffset? a, DateTimeOffset? b)
        => a == null ? b : b == null ? a : (a < b ? a : b);
}
